### ISSUE TITLE
spec: specify order of effects in x, y, z = f() assignment

### ISSUE THREAD

**Issue by alandonovan** [bazelbuild/starlark#150]
The spec does not mandate the order of effects in a compound assignment such as `x, y, z = f()`. The Java and Go  implementations, like Python, evaluate f() before the LHS expressions, as this is the only practical strategy. The spec should require it, and the implementations should have tests.

**Comment by brandjon** [comment]
Definitely, the rhs should be fully evaluated before any assignments take place.

We probably also have this problem elsewhere in the spec. For instance, we specify left-to-right evaluation for dictionary literals, and short-circuiting evaluation for boolean `or`/`and`. But what about binary operators like `+`? What about function call arguments, particularly when `*args` and `**kwargs` are involved? And what about compound assignments that involved nested tuple decomposition (`w, (x, y), z = f()`) -- is that depth-first (I think so) or breadth-first?

Ideally we'd remain compatible with whatever Python specifies while mandating an intuitive left-to-right order in all cases. If that's not satisfiable, we should still mandate the same order for all implementations.

Expanding this issue to encompass all those.
